crates/picky-krb/src/data_types.rs

```diff
@@ -152,28 +162,247 @@ pub type Ticket = ApplicationTag<TicketInner, 1>;
+///         cusec                   [4] Microseconds,
+///         ctime                   [5] KerberosTime,
+///         subkey                  [6] EncryptionKey OPTIONAL,
+///         seq-number              [7] UInt32 OPTIONAL,
+///         authorization-data      [8] AuthorizationData OPTIONAL
+/// }
+/// ```
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct AuthenticatorInner {
+    pub authenticator_bno: ExplicitContextTag0<IntegerAsn1>,
+    pub crealm: ExplicitContextTag1<Realm>,
+    pub cname: ExplicitContextTag2<PrincipalName>,
+    pub cksum: Optional<Option<ExplicitContextTag3<Checksum>>>,
+    pub cusec: ExplicitContextTag4<Microseconds>,
+    pub ctime: ExplicitContextTag5<KerberosTime>,
+    #[serde(default)]
+    pub subkey: Optional<Option<ExplicitContextTag6<EncryptionKey>>>,
+    #[serde(default)]
+    pub seq_number: Optional<Option<ExplicitContextTag7<IntegerAsn1>>>,
+    #[serde(default)]
+    pub authorization_data: Optional<Option<ExplicitContextTag8<AuthorizationData>>>,
+}
+pub type Authenticator = ApplicationTag<AuthenticatorInner, AUTHENTICATOR_TYPE>;
+
+/// [RFC 4120](https://datatracker.ietf.org/doc/html/rfc4120#section-5.5.2)
+///
+/// ```not_rust
+/// EncAPRepPart    ::= [APPLICATION 27] SEQUENCE {
+///         ctime           [0] KerberosTime,
+///         cusec           [1] Microseconds,
+///         subkey          [2] EncryptionKey OPTIONAL,
+///         seq-number      [3] UInt32 OPTIONAL
+/// }
+/// ```
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct EncApRepPartInner {
+    pub ctime: ExplicitContextTag0<KerberosTime>,
+    pub cusec: ExplicitContextTag1<Microseconds>,
+    #[serde(default)]
+    pub subkey: Optional<Option<ExplicitContextTag2<EncryptionKey>>>,
+    #[serde(default)]
+    pub seq_number: Optional<Option<ExplicitContextTag3<IntegerAsn1>>>,
+}
+pub type EncApRepPart = ApplicationTag<EncApRepPartInner, ENC_AP_REP_PART_TYPE>;
+
+/// [RFC 4120](https://datatracker.ietf.org/doc/html/rfc4120#section-5.2.7.5)
+///
+/// ```not_rust
+/// ETYPE-INFO2-ENTRY       ::= SEQUENCE {
+///         etype           [0] Int32,
+///         salt            [1] KerberosString OPTIONAL,
+///         s2kparams       [2] OCTET STRING OPTIONAL
+/// }
+/// ```
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct EtypeInfo2Entry {
+    pub etype: ExplicitContextTag0<IntegerAsn1>,
+    #[serde(default)]
+    pub salt: Optional<Option<ExplicitContextTag1<KerberosStringAsn1>>>,
+    #[serde(default)]
+    pub s2kparams: Optional<Option<ExplicitContextTag2<OctetStringAsn1>>>,
+}
+
+/// [RFC 4120](https://datatracker.ietf.org/doc/html/rfc4120#section-5.2.7.5)
+///
+/// ```not_rust
+/// ETYPE-INFO2              ::= SEQUENCE SIZE (1..MAX) OF ETYPE-INFO2-ENTRY
+/// ```
+pub type EtypeInfo2 = Asn1SequenceOf<EtypeInfo2Entry>;
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct KrbResult<T>(pub Result<T, KrbError>);
```

Comment:
Is it used outside of tests? What is the intended usage pattern?

As a general remark, I was expecting a `Result<T, KrbError>` instead of a newtype pattern.

I’m pretty sure you can implement `Deserialize` on it like done below on it, and get the benefits of using the `?` operator.

Reply:
А `KDC` response can be some valid data or encoded `KrbError`.  `KrbResult` is used for simplifying the decoding.

Reply:
Done